### PR TITLE
Ability to compute material volume fractions over mesh elements

### PR DESCRIPTION
**Pull request by paulromano**
# Description

This PR introduces a capability to compute the volume of individual materials within a mesh element. The intended purpose of this feature is to support mesh-based R2S workflows. Namely, this capability can be used to compute homogenized material compositions over mesh elements, which are then activated to create a decay photon source over a mesh. A few notes:

- I was able to reuse some of the logic from volume calculations by separating out a `reduce_indices_hits` function that is now used both for regular volume calculations as well as the new feature here.
- To support this feature, two new C API functions are added: `openmc_mesh_get_n_elements` and `openmc_mesh_volume_fractions`.
- Users can utilize this functionality through `openmc.lib` as the `Mesh.volume_fractions(...)` method.
- There was an indexing bug in `RectilinearMesh::volume` that is fixed here.
- Void volumes are calculated as well and are represented with `None` on the Python side

One of the challenging design aspects of this was how to handle memory allocation, since we don't know a priori how many materials are going to be hit in a given mesh element. The way I chose to handle this is as follows. First, the loop over mesh elements is done on the Python side. This avoids the need to handle a doubly nested allocation (akin to `<vector<vector<MaterialVolume>>`) and instead means we only need a single allocation per call to `openmc_mesh_volume_fractions`. This allocation is done on the Python side with ctypes with a guess on how many materials at most will be present (starts at 16). If you really run into a situation where a single element has _more_ than 16 materials, it will raise an exception on the Python side, which then triggers a reallocation with double the size.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: src/mesh.cpp [hunk 1]
```diff
@@ -1782,6 +1845,28 @@ extern "C" int openmc_mesh_set_id(int32_t index, int32_t id)
   return 0;
 }
 
+//! Get the number of elements in a mesh
+extern "C" int openmc_mesh_get_n_elements(int32_t index, size_t* n)
```
Comment (pshriwise) [inline src/mesh.cpp:1849]: For consistency we'll need entries in `capi.h` for these methods.

File: src/mesh.cpp [hunk 2]
```diff
@@ -142,6 +145,66 @@ vector<double> Mesh::volumes() const
...
+      }
+    }
+
+    // Reduce index/hits lists from each thread into a single copy
+    reduce_indices_hits(local_materials, local_hits, materials, hits);
+  }
+
+  // Make sure span passed in is large enough
+  if (hits.size() > result.size()) {
+    return -1;
```
Comment (pshriwise) [inline src/mesh.cpp:195]: To trigger the expected exception for analagous method in `openmc.lib` we'll want to return `OPENMC_E_ALLOCATE` here. To that end, a test with a contrived problem containing more than 16 materials in one of the mesh elements might be useful to exercise the code in `openmc.lib.mesh`.

File: include/openmc/volume_calc.h [hunk 3]
```diff
@@ -89,6 +92,47 @@ extern vector<VolumeCalculation> volume_calcs;
...
+template<typename T, typename T2>
+void reduce_indices_hits(const vector<T>& local_indices,
+  const vector<T2>& local_hits, vector<T>& indices, vector<T2>& hits)
+{
+#ifdef _OPENMP
+  int n_threads = omp_get_num_threads();
+#else
+  int n_threads = 1;
+#endif
+
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/volume_calc.h:110]: Gonna make a mini-PR moving this code into the `openmp_interface.h` file as this pattern appears in a few places now.

**Review by pshriwise — CHANGES_REQUESTED** [review]
Pleasantly surprised at the small amount of code this takes. Thanks @paulromano!! Not much to clean up here.

One design comment/thought that came to mind was that it might be useful to wrap the current version of `Mesh::volume_fractions` around a function that returns a vector of `MaterialVolumes` so we can suport return of dynamically sized results to support cases where this capability may be accessed via the C++ API.

File: src/mesh.cpp (hunk 2 above)
Comment (paulromano) [reply src/mesh.cpp:195]: This is exactly what happens in `openmc_mesh_volume_fractions` (if this method returns -1, `openmc_mesh_volume_fractions` returns `OPENMC_E_ALLOCATE`)

Comment (pshriwise) [reply src/mesh.cpp:195]: Ah right, my bad. I was interpreting the return value here as an errorcode (which it kind of is?), but this isn't the CAPI function obvi.

File: include/openmc/volume_calc.h [hunk 4]
```diff
@@ -89,6 +92,47 @@ extern vector<VolumeCalculation> volume_calcs;
...
+        }
+      }
+      if (!already_added) {
+        // If we made it here, the material hasn't yet been added to the
+        // master list, so add entries to the master indices and master
+        // hits lists
+        indices.push_back(local_indices[j]);
+        hits.push_back(local_hits[j]);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (paulromano) [reply include/openmc/volume_calc.h:132]: Nice simplification, thanks for the suggestion!

**Comment by paulromano** [comment]
@pshriwise All your feedback has been addressed. Note that I ended up changing the name of the method to `Mesh::material_volumes` since the method returns volumes, not volume fractions (returning volumes gives more information, and volume fractions can be calculated trivially after the fact). I also added a (currently unused) overload for `Mesh::material_volumes` that returns a `vector` as you had requested.

File: tests/unit_tests/test_lib.py [hunk 5]
```diff
@@ -580,6 +581,16 @@ def test_regular_mesh(lib_init):
     msf.translation = translation
     assert msf.translation == translation
 
+    # Test material volumes
+    mesh = openmc.lib.RegularMesh()
+    mesh.dimension = (2, 2, 1)
+    mesh.set_parameters(lower_left=(-0.63, -0.63, -0.5),
+                        upper_right=(0.63, 0.63, 0.5))
+    vol_fracs = mesh.material_volumes()
```
Comment (shimwell) [inline tests/unit_tests/test_lib.py:589]:
```suggestion
    vols = mesh.material_volumes()
```

File: tests/unit_tests/test_lib.py [hunk 6]
```diff
@@ -580,6 +581,16 @@ def test_regular_mesh(lib_init):
...
+    assert len(vol_fracs) == 4
+    for elem_fracs in vol_fracs:
```
Comment (shimwell) [inline tests/unit_tests/test_lib.py:591]:
```suggestion
    for elem_fracs in vols:
```

File: openmc/lib/mesh.py [hunk 7]
```diff
@@ -123,6 +142,67 @@ def id(self):
...
+
+        Returns
+        -------
+        List of tuple of (material, volume) for each mesh element. Void volume
+        is represented by having a value of None in the first element of a
+        tuple.
+
+        """
+        if n_samples <= 0:
+            raise ValueError("Number of samples must be positive")
```
Comment (shimwell) [inline openmc/lib/mesh.py:176]: should we perhaps add another check that checks the number of samples is above the number of mesh voxels and perhaps returns a warning

**Comment by shimwell** [comment]
Perhaps I've misunderstood something here but I tried to make a minimal example and was unsure about the result.

I thought the material mass would be 1/8 of the volume of the sphere as my mesh with 8 voxels covers the sphere, but I'm getting 1.0 for each material volume.

```python
import math
import openmc
import openmc.lib
mat_1 = openmc.Material(material_id=1) 
mat_1.add_element('Ag', 1)
mat_1.set_density('g/cm3', 1)
mat_2 = openmc.Material(material_id=2) 
mat_2.add_element('Au', 1)
mat_2.set_density('g/cm3', 1)
sphere_radius = 1
materials = openmc.Materials([mat_1, mat_2])
sph1 = openmc.Sphere(r=sphere_radius, boundary_type='vacuum')
zplane1 = openmc.ZPlane(z0=0)
reg1 = -sph1 & +zplane1
reg2 = -sph1 & -zplane1
cell1 = openmc.Cell(region=reg1)
cell2 = openmc.Cell(region=reg2)
cell1.fill = mat_1
cell2.fill = mat_2
geometry = openmc.Geometry([cell1, cell2])
settings = openmc.Settings()
settings.batches = 1
settings.particles = 1
settings.run_mode = 'fixed source'
model = openmc.Model(geometry=geometry, materials=materials, settings=settings)
model.export_to_model_xml()

openmc.lib.init()
mesh = openmc.lib.RegularMesh()
mesh.dimension = (2, 2, 2)  # 8 elements
mesh.set_parameters(
    lower_left=(-1, -1, -1), # fully covers the sphere of radius 1
    upper_right=(1, 1, 1),
)
volume_of_sphere = (4/3) * math.pi * math.pow(sphere_radius, 3)
vols = mesh.material_volumes(n_samples = 100000)
for vol in vols:
    print(vol)
    # assert vol[0][1] == volume_of_sphere/mesh.n_elements
```

File: openmc/lib/mesh.py (hunk 7 above)
Comment (paulromano) [reply openmc/lib/mesh.py:176]: This is the number of samples _per_ mesh element, so such a warning wouldn't make sense.

Comment (shimwell) [reply openmc/lib/mesh.py:176]: In that case should this number always be greater than 1?

Comment (paulromano) [reply openmc/lib/mesh.py:176]: It's up to the user to decide how many samples they think are necessary. If you had a mesh that was exactly conformal to material boundaries and had exactly one material per mesh element, using a single sample in each mesh element would yield the correct volumes. Of course, that's a pretty contrived example but just pointing out that really the only constraint (which is enforced here) is that the number of samples needs to be positive. Obviously for practical cases you would want to use a lot more than 1 sample per element to get accurate values.

**Comment by paulromano** [comment]
@shimwell Your example there helped to uncover a limitation, which was that we had implicitly assumed that the mesh was fully contained within the defined geometry. That limitation has now been removed. While looking at the surrounding code, I also realized that we were not handle random number seeds correctly in parallel, so that has also been fixed.

**Comment by shimwell** [comment]
Thanks for the fix Paul. Running the above script on the PR branch is now producing expected results.

I see there are tests in ```tests/unit_tests/test_lib.py`` that check the material volumes for a few situations.

I can see a tests where the material volumes are 100% of the mesh cell.
I can see a tests where there is a single material in the mesh voxel.

Wondering if it is worth adding a test where there are 2 materials and they don't fill the voxel 100%

**Comment by paulromano** [comment]
@shimwell This snippet covers the case you just mentioned (multiple materials that don't fill a voxel 100%):

https://github.com/paulromano/openmc/blob/0edd9ba29b82b088ad84f0df461e8e2e7b0b0a16/tests/unit_tests/test_lib.py#L594-L601

File: tests/unit_tests/test_lib.py [hunk 8]
```diff
@@ -580,6 +581,25 @@ def test_regular_mesh(lib_init):
...
+    mesh.dimension = (2, 2, 1)
+    mesh.set_parameters(lower_left=(-0.63, -0.63, -0.5),
+                        upper_right=(0.63, 0.63, 0.5))
+    vols = mesh.material_volumes()
+    assert len(vols) == 4
+    for elem_vols in vols:
+        assert sum(f[1] for f in elem_vols) == pytest.approx(1.26 * 1.26 / 4)
+
+    # If the mesh extends beyond the boundaries of the model, the volumes should
+    # still be reported correctted
```
Comment (shimwell) [inline tests/unit_tests/test_lib.py:595]:
```suggestion
    # still be reported corrected
```

Comment (paulromano) [reply tests/unit_tests/test_lib.py]: Oops, fixed!

**Comment by shimwell** [comment]
This looks good to me. I'm now wondering how you make use of this in the r2S workflow. Perhaps best if @pshriwise

**Comment by paulromano** [comment]
I'll have some mesh-based R2S examples to share in the near-term that use this capability.

@pshriwise Any further requests on this PR?

**Review by pshriwise — APPROVED** [review]
No further comments from me! I'll give this until tomorrow morning as it's a substantial feature addition. Looking forward to trying it out @paulromano! Thanks!

**Pull request merged**